packages/go_router/lib/src/route.dart

```diff
@@ -320,6 +325,49 @@ class GoRoute extends RouteBase {
   late final RegExp _pathRE;
 }
 
+/// Base class for classes that act as shells for sub-routes, such
+/// as [ShellRoute] and [StatefulShellRoute].
+abstract class ShellRouteBase extends RouteBase {
+  /// Constructs a [ShellRouteBase].
+  const ShellRouteBase._({super.routes}) : super._();
+
+  /// Attempts to build the Widget representing this shell route.
+  ///
+  /// Returns null if this shell route does not build a Widget, but instead uses
+  /// a Page to represent itself (see [buildPage]).
+  Widget? buildWidget(BuildContext context, GoRouterState state,
+      ShellNavigatorBuilder navigatorBuilder);
+
+  /// Attempts to build the Page representing this shell route.
+  ///
+  /// Returns null if this shell route does not build a Page, but instead uses
+  /// a Widget to represent itself (see [buildWidget]).
+  Page<dynamic>? buildPage(BuildContext context, GoRouterState state,
+      ShellNavigatorBuilder navigatorBuilder);
+
+  /// Returns the key for the [Navigator] that is to be used for the specified
+  /// immediate sub-route of this shell route.
+  GlobalKey<NavigatorState> navigatorKeyForSubRoute(RouteBase subRoute);
+}
+
+/// Navigator builder for shell routes.
+abstract class ShellNavigatorBuilder {
+  /// The [GlobalKey] to be used by the [Navigator] built for the current route.
+  GlobalKey<NavigatorState> get navigatorKey;
+
+  /// The route match list for the current route.
+  RouteMatchList get routeMatchList;
+
+  /// The route builder.
+  RouteBuilder get routeBuilder;
```

Comment:
This seems out of place. I think this is for preloading. Even so, this property should not be passed through this class. Let's remove it for now if we can postpone the preloading logic

Reply:
Yes, RouteBuilder is needed for preloading. You could alternatively get it via GoRouterDelegate I guess, by exposing that field there (and I did at some point), but we can see what work work best going forward (in a separate PR). Will remove the field for now.